lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/WordStorage.java

```diff
@@ -0,0 +1,338 @@
+
+    private byte[] wordData;
+
+    private int commonPrefixLength, commonPrefixPos;
+    private String currentEntry = null;
+    private final List<char[]> group = new ArrayList<>();
+    private final List<Integer> morphDataIDs = new ArrayList<>();
+
+    Builder(int wordCount, boolean hasCustomMorphData, FlagEnumerator flagEnumerator) {
+      this.flagEnumerator = flagEnumerator;
+      this.hasCustomMorphData = hasCustomMorphData;
+
+      hashTable = new int[wordCount];
+      wordData = new byte[wordCount * 6];
+
+      dataWriter = new ByteArrayDataOutput(wordData);
+      dataWriter.writeByte((byte) 0); // zero index is root, contains nothing
+      chainLengths = new int[hashTable.length];
+    }
+
+    void add(String entry, char[] flags, int morphDataID) throws IOException {
+      if (!entry.equals(currentEntry)) {
+        if (currentEntry != null) {
+          if (entry.compareTo(currentEntry) < 0) {
+            throw new IllegalArgumentException("out of order: " + entry + " < " + currentEntry);
+          }
+          int pos = flushGroup();
+
+          commonPrefixLength = GeneratingSuggester.commonPrefix(currentEntry, entry);
+          ByteArrayDataInput in = new ByteArrayDataInput(wordData);
+          in.setPosition(pos);
+          for (int i = currentEntry.length() - 1; i >= commonPrefixLength; i--) {
+            char c = (char) in.readVInt();
+            assert c == currentEntry.charAt(i);
+            pos -= in.readVInt();
+            in.setPosition(pos);
+          }
+          commonPrefixPos = pos;
+        }
+        currentEntry = entry;
+      }
+
+      group.add(flags);
+      if (hasCustomMorphData) {
+        morphDataIDs.add(morphDataID);
+      }
+    }
+
+    private int flushGroup() throws IOException {
+      IntsRefBuilder currentOrds = new IntsRefBuilder();
+
+      boolean hasNonHidden = false;
+      for (char[] flags : group) {
+        if (!hasHiddenFlag(flags)) {
+          hasNonHidden = true;
+          break;
+        }
+      }
+
+      for (int i = 0; i < group.size(); i++) {
+        char[] flags = group.get(i);
+        if (hasNonHidden && hasHiddenFlag(flags)) {
+          continue;
+        }
+
+        currentOrds.append(flagEnumerator.add(flags));
+        if (hasCustomMorphData) {
+          currentOrds.append(morphDataIDs.get(i));
+        }
+      }
+
+      int lastPos = commonPrefixPos;
+      for (int i = commonPrefixLength; i < currentEntry.length() - 1; i++) {
+        int pos = dataWriter.getPosition();
+        ensureArraySize(0, false);
+        dataWriter.writeVInt(currentEntry.charAt(i));
+        dataWriter.writeVInt(pos - lastPos);
+        lastPos = pos;
+      }
+
+      int pos = dataWriter.getPosition();
+      int hash = Math.abs(currentEntry.hashCode() % hashTable.length);
+      int collision = hashTable[hash];
+      hashTable[hash] = collision == 0 ? pos : -pos;
+
+      if (++chainLengths[hash] > 20) {
+        throw new RuntimeException(
```

Comment:
Awesome!  I am glad we are hard-rejecting accidentally adversarial uses.

But, isn't this ... statistically a small but non-trivial possibility?  Hashing is hard ... every fixed hash function has many adversaries, but perhaps not that common in real-world usage?

Anyway +1 to keep this safety.

Hmm should we maybe also confirm that the number of `.add(...)` calls does not exceed the original `int wordCount`?  That might otherwise cause excessive conflicts ...

Reply:
Yes, it's a nontrivial possibility. I thought following the pattern `please report this to dev@lucene.apache.org` would be enough (in which case we could either raise this limit, or rehash automatically, or suggest to alter the dictionary if it's wrong). Do you think we should be more proactive here?

Call counting looks good, will do!